build-tools/packages/build-cli/src/commands/generate/typetests.ts

```diff
@@ -0,0 +1,371 @@
+	type TypeData,
+	buildTestCase,
+	getFullTypeName,
+	getNodeTypeData,
+	typeOnly,
+} from "@fluidframework/build-tools";
+import { Flags } from "@oclif/core";
+import { PackageName } from "@rushstack/node-core-library";
+import * as changeCase from "change-case";
+import { mkdirSync, readJson, rmSync, writeFileSync } from "fs-extra";
+import * as resolve from "resolve.exports";
+import { ModuleKind, ModuleResolutionKind, Project, type SourceFile } from "ts-morph";
+import { PackageCommand } from "../../BasePackageCommand";
+import { ApiLevel, ensureDevDependencyExists, knownApiLevels } from "../../library";
+import { unscopedPackageNameString } from "./entrypoints";
+
+export default class GenerateTypetestsCommand extends PackageCommand<
+	typeof GenerateTypetestsCommand
+> {
+	static readonly description = "Generates type tests for a package or group of packages.";
+
+	static readonly flags = {
+		level: Flags.string({
+			description: "What API level to generate tests for.",
+			default: ApiLevel.internal,
+			options: knownApiLevels,
+		}),
+		outDir: Flags.directory({
+			description: "Where to emit the type tests file.",
+			default: "./src/test/types",
+		}),
+		outFile: Flags.string({
+			description: `File name for the generated type tests. The pattern '${unscopedPackageNameString}' within the value will be replaced with the unscoped name of this package in PascalCase.`,
+			default: `validate${unscopedPackageNameString}Previous.generated.ts`,
+		}),
+		publicFallback: Flags.boolean({
+			description:
+				"Use the public entrypoint as a fallback if the API at the requested level is not found.",
+			default: false,
+		}),
+		...PackageCommand.flags,
+	} as const;
+
+	protected async processPackage(pkg: Package): Promise<void> {
+		const { outDir, outFile } = this.flags;
+
+		// This cast is safe because oclif has already ensured only known ApiLevel values get to this point.
+		const level = this.flags.level as ApiLevel;
+		const fallbackLevel = this.flags.publicFallback ? ApiLevel.public : undefined;
+
+		// Do not check that file exists before opening:
+		// Doing so is a time of use vs time of check issue so opening the file could fail anyway.
+		// Do not catch error from opening file since the default behavior is fine (exits process with error showing useful message)
+		const currentPackageJson = pkg.packageJson;
+		const previousPackageName = `${currentPackageJson.name}-previous`;
+		const previousBasePath = path.join(pkg.directory, "node_modules", previousPackageName);
+		const previousPackageJsonPath = path.join(previousBasePath, "package.json");
+
+		ensureDevDependencyExists(currentPackageJson, previousPackageName);
+		this.verbose(`Reading package.json at ${previousPackageJsonPath}`);
+		const previousPackageJson = (await readJson(previousPackageJsonPath)) as PackageJson;
+		// Set the name in the JSON to the calculated previous package name, since the name in the previous package.json is
+		// the same as current.
```

Comment:
I think that's the only effect of this?
```suggestion
		// Set the name in the JSON to the calculated previous package name, since the name in the previous package.json is
		// the same as current. Just for making logs clearer.
```

Reply:
No, it ends up being used in a couple of places. One is the generated type test imports - it needs to know the rewritten name of the package at that point, and it's easier to set it here than to pass along the "real name" everywhere. I can clarify that in the comments.